File: ios/Sources/BannerManager.swift
// Copyright 2026 tauri-plugin-ad2mob contributors
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

import Foundation
import GoogleMobileAds
import Tauri
import UIKit

/// Lifecycle of the banner slot: creation, constraint-based insertion into the
/// Tauri root view (respecting safe areas: notch / Dynamic Island / home
/// indicator), position changes, hide and teardown.
///
/// Threading: every command from `AdmobPlugin` arrives on the main queue and
/// every SDK callback hops to the main queue before touching state, so all
/// fields are main-queue-confined. The `showBanner` invoke is settled exactly
/// once: resolved when the SDK reports the ad and the view was revealed,
/// rejected with `LOAD_FAILED` otherwise (a failed banner never stays
/// attached to the view hierarchy).
final class BannerManager: NSObject, GADBannerViewDelegate {
  private weak var plugin: AdmobPlugin?
  /// The live banner view, if any. A new `show` replaces it.
  private var bannerView: GADBannerView?
  /// Active Auto Layout constraints, deactivated/rebuilt on position changes.
  private var bannerConstraints: [NSLayoutConstraint] = []
  /// Ad unit of the current banner, used for event payloads.
  private var adUnitId: String?
  /// The `showBanner` invoke, settled exactly once on load success or failure.
  private var pendingShowInvoke: Invoke?

  init(plugin: AdmobPlugin) {
    self.plugin = plugin
    super.init()
  }

  /// Creates (replacing any previous banner), loads and attaches the banner.
  /// Inputs are already validated and parsed by the plugin. Called on the
  /// main queue.
  func show(
    adUnitId: String,
    position: AdmobMappers.BannerPositionDescriptor,
    size: AdmobMappers.BannerSizeDescriptor,
    invoke: Invoke
  ) {
    guard let container = plugin?.containerView() else {
      invoke.reject("no container view available to attach the banner", code: "NATIVE_ERROR")
      return
    }

    // One banner slot at a time: settle a superseded pending show and tear the
    // previous banner down before building the new one.
    if let pending = pendingShowInvoke {
      pendingShowInvoke = nil
      pending.reject("showBanner superseded by a new showBanner call", code: "LOAD_FAILED")
    }
    teardownBannerView()

    self.adUnitId = adUnitId
    pendingShowInvoke = invoke

    let adSize = resolvedAdSize(for: size, container: container)
    let banner = GADBannerView(adSize: adSize)
    banner.translatesAutoresizingMaskIntoConstraints = false
    banner.adUnitID = adUnitId
    banner.rootViewController = plugin?.rootViewController()
    banner.delegate = self
    banner.isHidden = true  // revealed when the SDK reports the ad
    container.addSubview(banner)
    bannerView = banner
    applyConstraints(for: banner, in: container, position: position, height: adSize.size.height)

    plugin?.logDebug("banner loading (\(adUnitId), \(position.rawValue), \(size))")
    banner.load(GADRequest())
  }

  /// Hides the banner while keeping its constraints for a later re-show.
  /// No-op when there is no banner. Called on the main queue.
  func hide() {
    bannerView?.isHidden = true
  }

  /// Rebuilds the placement constraints of a live banner. No-op when there is
  /// none (the caller still resolves). Called on the main queue.
  func setPosition(_ position: String) {
    guard let parsed = AdmobMappers.parseBannerPosition(position),
      let banner = bannerView,
      let container = banner.superview
    else {
      return
    }
    NSLayoutConstraint.deactivate(bannerConstraints)
    applyConstraints(for: banner, in: container, position: parsed, height: banner.adSize.size.height)
  }

  /// Removes the banner from the hierarchy and clears every reference. Safe
  /// when nothing was created. Called on the main queue.
  func destroy() {
    teardownBannerView()
    adUnitId = nil
    if let pending = pendingShowInvoke {
      pendingShowInvoke = nil
      pending.reject("banner destroyed while loading", code: "LOAD_FAILED")
    }
  }

  // MARK: - GADBannerViewDelegate

  func bannerViewDidReceiveAd(_ bannerView: GADBannerView) {
    DispatchQueue.main.async { [weak self] in
      guard let self, bannerView === self.bannerView else { return }
      bannerView.isHidden = false  // reveal
      self.plugin?.logDebug("banner loaded")
      self.plugin?.trigger(AdmobEvents.BANNER_LOADED, data: self.eventPayload())
      self.pendingShowInvoke?.resolve()
      self.pendingShowInvoke = nil
    }
  }

  func bannerView(_ bannerView: GADBannerView, didFailToReceiveAdWithError error: Error) {
    DispatchQueue.main.async { [weak self] in
      guard let self, bannerView === self.bannerView else { return }
      let nsError = error as NSError
      var payload = self.eventPayload()
      payload["code"] = "\(nsError.code)"
      payload["message"] = error.localizedDescription
      payload["domain"] = nsError.domain
      self.plugin?.trigger(AdmobEvents.BANNER_FAILED, data: payload)
      self.pendingShowInvoke?.reject(
        "banner failed to load: \(error.localizedDescription)", code: "LOAD_FAILED")
      self.pendingShowInvoke = nil
      // Do not keep a failed banner attached to the view hierarchy.
      self.teardownBannerView()
      self.adUnitId = nil
    }
  }

  func bannerViewDidRecordImpression(_ bannerView: GADBannerView) {
    DispatchQueue.main.async { [weak self] in
      guard let self, bannerView === self.bannerView else { return }
      self.plugin?.trigger(AdmobEvents.BANNER_IMPRESSION, data: self.eventPayload())
    }
  }

  func bannerViewDidRecordClick(_ bannerView: GADBannerView) {
    DispatchQueue.main.async { [weak self] in
      guard let self, bannerView === self.bannerView else { return }
      self.plugin?.trigger(AdmobEvents.BANNER_CLICKED, data: self.eventPayload())
    }
  }

  func bannerViewWillPresentScreen(_ bannerView: GADBannerView) {
    DispatchQueue.main.async { [weak self] in
      guard let self, bannerView === self.bannerView else { return }
      self.plugin?.trigger(AdmobEvents.BANNER_OPENED, data: self.eventPayload())
    }
  }

  func bannerViewDidDismissScreen(_ bannerView: GADBannerView) {
    DispatchQueue.main.async { [weak self] in
      guard let self, bannerView === self.bannerView else { return }
      self.plugin?.trigger(AdmobEvents.BANNER_CLOSED, data: self.eventPayload())
    }
  }

  // MARK: - Private

  /// Maps a parsed size descriptor to the concrete `GADAdSize`. Adaptive
  /// banners are anchored to the current orientation and sized from the
  /// container width.
  private func resolvedAdSize(
    for descriptor: AdmobMappers.BannerSizeDescriptor,
    container: UIView
  ) -> GADAdSize {
    switch descriptor {
    case .banner:
      return GADAdSizeBanner
    case .largeBanner:
      return GADAdSizeLargeBanner
    case .mediumRectangle:
      return GADAdSizeMediumRectangle
    case .fullBanner:
      return GADAdSizeFullBanner
    case .leaderboard:
      return GADAdSizeLeaderboard
    case .adaptive:
      return GADCurrentOrientationAnchoredAdaptiveBannerAdSizeWithWidth(adaptiveWidth(for: container))
    }
  }

  /// Width used for anchored adaptive banners: the container width when it is
  /// laid out, otherwise the screen width of the hosting window.
  private func adaptiveWidth(for container: UIView) -> CGFloat {
    if container.bounds.width > 0 {
      return container.bounds.width
    }
    if let window = container.window, let scene = window.windowScene {
      return scene.screen.bounds.width
    }
    return UIScreen.main.bounds.width
  }

  /// Pins the banner to the leading/trailing edges and the requested edge of
  /// the container's `safeAreaLayoutGuide`, so notch / Dynamic Island / home
  /// indicator are respected.
  private func applyConstraints(
    for banner: GADBannerView,
    in container: UIView,
    position: AdmobMappers.BannerPositionDescriptor,
    height: CGFloat
  ) {
    var constraints: [NSLayoutConstraint] = [
      banner.leadingAnchor.constraint(equalTo: container.leadingAnchor),
      banner.trailingAnchor.constraint(equalTo: container.trailingAnchor),
      banner.heightAnchor.constraint(equalToConstant: height),
    ]
    switch position {
    case .bottom:
      constraints.append(
        banner.bottomAnchor.constraint(equalTo: container.safeAreaLayoutGuide.bottomAnchor))
    case .top:
      constraints.append(
        banner.topAnchor.constraint(equalTo: container.safeAreaLayoutGuide.topAnchor))
    }
    NSLayoutConstraint.activate(constraints)
    bannerConstraints = constraints
  }

  /// Detaches the current banner view: nils the delegate, deactivates its
  /// constraints and removes it from the superview.
  private func teardownBannerView() {
    NSLayoutConstraint.deactivate(bannerConstraints)
    bannerConstraints = []
    if let banner = bannerView {
      banner.delegate = nil
      banner.removeFromSuperview()
    }
    bannerView = nil
  }

  /// Base payload for banner events: the current ad unit.
  private func eventPayload() -> JSObject {
    var payload: JSObject = [:]
    if let adUnitId {
      payload["adUnitId"] = adUnitId
    }
    return payload
  }
}

File: android/src/main/java/com/plugin/admob/BannerManager.kt
// SPDX-License-Identifier: Apache-2.0
// SPDX-License-Identifier: MIT

package com.plugin.admob

import android.app.Activity
import android.content.res.Resources
import android.graphics.Color
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.AdSize
import com.google.android.gms.ads.AdView
import com.google.android.gms.ads.LoadAdError

/**
 * Owns the native banner: creates the `AdView`, wraps it in a container
 * `FrameLayout` inserted into the activity content root (the same root that
 * hosts the Tauri WebView), applies safe-area padding and reports SDK events.
 *
 * All methods must be called from the UI thread (the AdMob SDK delivers its
 * callbacks on the main thread, and the plugin hops to the UI thread before
 * entering this class), so no synchronization is required.
 *
 * Safe areas: for `bottom` banners the navigation-bar bottom inset is applied
 * as padding so edge-to-edge mode never covers the ad; for `top` banners the
 * status-bar top inset is applied instead. Insets are read through
 * `ViewCompat.setOnApplyWindowInsetsListener` and re-applied live when the
 * position changes.
 *
 * @param activity the foreground activity whose content root hosts the banner
 * @param notifyEvent called with (event name, payload) for every SDK event
 */
internal class BannerManager(
    private val activity: Activity,
    private val notifyEvent: (event: String, payload: JSObject) -> Unit
) {
    private var container: FrameLayout? = null
    private var adView: AdView? = null
    private var position: BannerPosition = BannerPosition.BOTTOM

    /**
     * `true` while the user hid the banner through `hideBanner`. A load that
     * completes while hidden must not reveal the view again.
     */
    private var isHidden = false

    /**
     * Bumped on every teardown ([destroyView]). Load callbacks created with
     * an older epoch are stale (superseded by a newer `showBanner` or a
     * destroy) and complete their own invoke exactly once without touching
     * the current state.
     */
    private var loadEpoch = 0L

    /** Last insets dispatched to the container, re-applied on position change. */
    private var lastInsets: WindowInsetsCompat? = null

    /**
     * Android-only thin wrapper that turns a parsed [BannerSize] into a
     * concrete `AdSize`. Fixed sizes map to the SDK constants; `ADAPTIVE` is
     * computed from the current screen width in dp for the current
     * orientation. Returns `null` when the SDK cannot compute the adaptive
     * size (screen too narrow), which the caller reports as
     * `INVALID_ARGUMENT`.
     */
    private fun resolveAdSize(size: BannerSize): AdSize? = when (size) {
        BannerSize.BANNER -> AdSize.BANNER
        BannerSize.LARGE_BANNER -> AdSize.LARGE_BANNER
        BannerSize.MEDIUM_RECTANGLE -> AdSize.MEDIUM_RECTANGLE
        BannerSize.FULL_BANNER -> AdSize.FULL_BANNER
        BannerSize.LEADERBOARD -> AdSize.LEADERBOARD
        BannerSize.ADAPTIVE -> {
            val metrics = Resources.getSystem().displayMetrics
            val widthDp = (metrics.widthPixels / metrics.density).toInt()
            AdSize.getCurrentOrientationAnchoredAdaptiveBannerAdSize(activity, widthDp)
        }
    }

    /**
     * Creates a fresh banner for the given arguments and starts loading it.
     * Any previous banner is destroyed first (a new `showBanner` always
     * replaces the old one). The container is attached hidden and revealed
     * only when the ad loads, so a failed load never leaves anything visible.
     *
     * The [Invoke] is resolved/rejected exactly once — from the load
     * callbacks of THIS banner (stale callbacks from a superseded banner
     * resolve their own invoke silently).
     */
    fun show(adUnitId: String, newPosition: BannerPosition, size: BannerSize, invoke: Invoke) {
        val adSize = resolveAdSize(size)
        if (adSize == null) {
            invoke.reject(
                "unable to compute the adaptive banner size for the current screen",
                AdmobRejectCodes.INVALID_ARGUMENT
            )
            return
        }

        // destroyView() already bumps the epoch, invalidating the callbacks
        // of the previous banner (if any).
        destroyView()
        val epoch = loadEpoch
        position = newPosition
        isHidden = false

        val newContainer = FrameLayout(activity)
        newContainer.setBackgroundColor(Color.TRANSPARENT)

        val newAdView = AdView(activity)
        newAdView.setAdSize(adSize)
        newAdView.adUnitId = adUnitId
        newContainer.addView(
            newAdView,
            FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT
            )
        )

        container = newContainer
        adView = newAdView

        ViewCompat.setOnApplyWindowInsetsListener(newContainer) { _, insets ->
            lastInsets = insets
            applyInsetsPadding()
            insets
        }

        newAdView.adListener = object : AdListener() {
            override fun onAdLoaded() {
                if (epoch != loadEpoch) {
                    // Superseded by a newer showBanner/destroy: this command
                    // still needs its response, but the state must not change.
                    invoke.resolve()
                    return
                }
                newContainer.visibility = if (isHidden) View.GONE else View.VISIBLE
                notifyEvent(AdmobEvents.BANNER_LOADED, JSObject().put("adUnitId", adUnitId))
                invoke.resolve()
            }

            override fun onAdFailedToLoad(loadError: LoadAdError) {
                if (epoch != loadEpoch) {
                    invoke.resolve()
                    return
                }
                destroyView()
                notifyEvent(
                    AdmobEvents.BANNER_FAILED,
                    JSObject()
                        .put("adUnitId", adUnitId)
                        .put("code", loadError.code.toString())
                        .put("message", loadError.message)
                        .put("domain", loadError.domain)
                )
                invoke.reject(
                    "banner load failed: ${loadError.message}",
                    AdmobRejectCodes.LOAD_FAILED
                )
            }

            override fun onAdOpened() {
                notifyEvent(AdmobEvents.BANNER_OPENED, JSObject().put("adUnitId", adUnitId))
            }

            override fun onAdClicked() {
                notifyEvent(AdmobEvents.BANNER_CLICKED, JSObject().put("adUnitId", adUnitId))
            }

            override fun onAdImpression() {
                notifyEvent(AdmobEvents.BANNER_IMPRESSION, JSObject().put("adUnitId", adUnitId))
            }

            override fun onAdClosed() {
                notifyEvent(AdmobEvents.BANNER_CLOSED, JSObject().put("adUnitId", adUnitId))
            }
        }

        val contentRoot = activity.findViewById<ViewGroup>(android.R.id.content)
        contentRoot.addView(newContainer, makeLayoutParams())
        applyInsetsPadding()
        newContainer.visibility = View.GONE

        newAdView.loadAd(AdRequest.Builder().build())
    }

    /**
     * Hides the banner while keeping the view alive for a re-show. When no
     * banner exists this is a no-op (the contract resolves unconditionally).
     */
    fun hide() {
        isHidden = true
        container?.visibility = View.GONE
    }

    /**
     * Repositions the live banner (gravity + safe-area padding). No-op when
     * no banner exists.
     */
    fun setPosition(newPosition: BannerPosition) {
        position = newPosition
        val currentContainer = container ?: return
        val params = currentContainer.layoutParams as? FrameLayout.LayoutParams ?: return
        params.gravity = gravityFor(position)
        currentContainer.layoutParams = params
        applyInsetsPadding()
    }

    /**
     * Removes the banner from the hierarchy and destroys the `AdView`. Any
     * in-flight load callback becomes stale and completes its own invoke
     * without touching state.
     */
    fun destroy() {
        destroyView()
    }

    private fun gravityFor(position: BannerPosition): Int = when (position) {
        BannerPosition.TOP -> Gravity.TOP or Gravity.CENTER_HORIZONTAL
        BannerPosition.BOTTOM -> Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
    }

    private fun makeLayoutParams(): FrameLayout.LayoutParams {
        val params = FrameLayout.LayoutParams(
            FrameLayout.LayoutParams.WRAP_CONTENT,
            FrameLayout.LayoutParams.WRAP_CONTENT
        )
        params.gravity = gravityFor(position)
        return params
    }

    /**
     * Applies the safe-area padding for the current position using the last
     * dispatched insets: status-bar top inset for `top` banners,
     * navigation-bar bottom inset for `bottom` banners (edge-to-edge safety).
     */
    private fun applyInsetsPadding() {
        val currentContainer = container ?: return
        val insets = lastInsets ?: return
        val statusBarTop = insets.getInsets(WindowInsetsCompat.Type.statusBars()).top
        val navigationBarBottom = insets.getInsets(WindowInsetsCompat.Type.navigationBars()).bottom
        when (position) {
            BannerPosition.TOP -> currentContainer.setPadding(0, statusBarTop, 0, 0)
            BannerPosition.BOTTOM -> currentContainer.setPadding(0, 0, 0, navigationBarBottom)
        }
    }

    /** Tears down the current banner (if any) and bumps the staleness epoch. */
    private fun destroyView() {
        loadEpoch += 1
        val currentContainer = container
        val currentAdView = adView
        container = null
        adView = null
        isHidden = false
        lastInsets = null
        if (currentAdView != null) {
            currentAdView.destroy()
        }
        if (currentContainer != null) {
            (currentContainer.parent as? ViewGroup)?.removeView(currentContainer)
        }
    }
}
